drivers/pca95xx/pca95xx.c

```diff
@@ -0,0 +1,486 @@
+/*
+ * Copyright (C) 2018 Matthew Blue <matthew.blue.neuro@gmail.com>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_pca95xx
+ * @{
+ *
+ * @file
+ * @brief       PCA95xx I2C GPIO expander device driver
+ *
+ * @author      Matthew Blue <matthew.blue.neuro@gmail.com>
+ * @}
+ */
+
+#include "assert.h"
+#include "periph/i2c.h"
+#include "periph/gpio.h"
+
+#include "pca95xx.h"
+#include "pca95xx_params.h"
+#include "pca95xx_regs.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+#define I2C_SPEED I2C_SPEED_FAST
+
+#define I2C (dev->params.i2c)
+#define ADDR (dev->params.addr)
+#define DFLAGS (dev->params.dflags)
+#define PFLAGS (dev->pflags)
+
+static inline uint8_t _get_reg(uint8_t pin, pca95xx_dflags_t flags, uint8_t reg)
+{
+    /* Devices with >8 bits have twice as many registers */
+    if (flags & (1 << PCA95XX_16PIN_DEV)) {
+        reg <<= 1;
+    }
+
+    /* Offset reg for high pins */
+    if (pin >= 8) {
+        reg++;
+    }
+
+    return reg;
+}
+
+static inline uint8_t _pin_bit(uint8_t pin)
+{
+    if (pin < 8) {
+        return pin;
+    }
+    else {
+        return pin - 8;
+    }
+}
+
+static inline uint8_t _get_pflag(uint8_t pin, uint32_t pflags, uint8_t flag)
+{
+    /* Shift to the correct entry */
+    pflags >>= 2 * pin;
+
+    if (pflags & (1 << flag)) {
+        return 1;
+    }
+    else {
+        return 0;
+    }
+}
+
+static inline void _set_pflag(uint8_t pin, uint32_t *pflags,
+                              uint8_t flag, uint8_t val)
+{
+    if (val) {
+        *pflags |= (1UL << (2 * pin + flag));
+    }
+    else {
+        *pflags &= ~(1UL << (2 * pin + flag));
+    }
+}
+
+static inline uint8_t _get_int_pin(uint8_t int_info)
+{
+    /* INT pin is lowest nibble of int_info */
+    return (int_info & 0x0F);
+}
+
+static inline int8_t _get_int_num(uint8_t pin, uint8_t int_info[])
+{
+    uint8_t num;
+
+    for (num = 0; num < PCA95XX_MAX_INTS; num++) {
+        if (_get_int_pin(int_info[num]) == pin) {
+            return num;
+        }
+    }
+
+    /* No entry with that pin */
+    return -1;
+}
+
+static void _isr_cb_mux(void *arg)
+{
+    pca95xx_t *dev = (pca95xx_t *)arg;
+    uint8_t reg_addr;
+    uint8_t reg[2];
+    uint8_t num;
+    uint8_t pin;
+    uint8_t pin_val;
+
+    i2c_acquire(I2C);
+
+    /* Get input register */
+    reg_addr = _get_reg(0, DFLAGS, PCA95XX_INPUT_ADDR);
+    i2c_read_reg(I2C, ADDR, reg_addr, &reg[0]);
```

Comment:
As I already explained in PR #10430, this I2C read access might collide with an I2C write access to an output port in the case where ports of different expander devices at the same I2C bus are interconnected. This is at least the case for PCF857x I/O expander.

The reason is that an I2C write access to an output port of one device may cause an interrupt by an input port of another device before the I2C write access to the first device has been finished. The handling of the interrupt of the second device requires I2C read access. If the devices are connected to the same I2C bus, the current I2C write access to the first device avoids the required I2C read access from interrupt handler to the second device. The I2C read access fails and the interrupt is lost.

Because the interrupt handling occurs in the same thread context as the I2C write access which is interrupted, the mutex-based `i2c_aquire` does not help to prevent this conflict.
Even though `mutex_lock` sees a mutex that is locked it returns with continuation of the thread that has locked the mutex. Unfortunatly, this is the same thread that is in interrupt handling now :worried: That is, it starts to try the I2C read access even though the I2C bus is occupied.

Reply:
Interesting problem! Let's find a solution in #10430, so that we are not splitting the discussion. I will then apply a similar fix here.